Add support for SASL SCRAM

GODIMONGO opened the pull request.

amotl commented:
Hi @GODIMONGO. Thanks for your contribution, and a happy new year. Can you address a few admonitions by Copilot?
@normanr, @Neustradamus, and potentially others: Can you add a few review cycles here to evaluate if the patch is good to go?

Neustradamus commented:
@GODIMONGO: Nice PR, it is an important feature.

Linked to:
- https://github.com/xmpppy/xmpppy/issues/32
- https://github.com/scram-sasl/info/issues/1

Can you check to have a perfect PR?

Thanks a lot in advance.

GODIMONGO commented:
Yeah, good, I didn't think there would be such a buzz around this feature. I already shut down the server, but it looks like I'll have to start it up again. I'll do it soon, though.

GODIMONGO commented:
@Neustradamus
I fixed the errors and tested it on my server. Authentication works and messages are being sent!

normanr commented:
Some general comments:
- Please could you rebase the commits on top of the current HEAD to remove the merge commits
- Please write the commit messages in English
- Please address the Copilot suggestions (especially the ones for insecure code and error handling, but spelling mistakes are also important to fix)

Stylistically I think it would be nice to abstract out a "SASL mechanism" class (like the old patch I linked to in [this comment](https://github.com/xmpppy/xmpppy/issues/32#issuecomment-3609599212)). I don't think the support for all SASL mechanisms should all be hardcoded directly into the `SASL` class.

You could maybe have a common base class for mechanisms supported by scramp, and then (if needed) have a subclass for each of the individual mechanisms. It certainly wouldn't need to support the dynamic importing that the patch has, the classes could just be registered (maybe plugged in?) with the `SASL` class so that it can find them by looking them up in a dictionary by mechanism name.

normanr commented on auth.py:
```diff
@@ -0,0 +1,467 @@
+        self.username=username
+        self.password=password
+        self.mechanism=None
+        self.scram_state={}
```
Move the state to ScramBase

normanr commented on scram.py:
```diff
@@ -0,0 +1,126 @@
+        self.sasl.scram_state = {}
+        nonce_bytes = __import__('os').urandom(18)
+        nonce = base64.b64encode(nonce_bytes).decode('ascii')
+        gs2_header = self.sasl._scram_build_gs2(cb_type if mechanism.endswith('PLUS') else None)
```
I think it would be slightly nice style to introduce another 2 classes and make the "Plusness" a class constant instead of checking the mechanism name suffix in this method.

normanr commented on auth.py:
```diff
@@ -0,0 +1,467 @@
+    def _scram_build_gs2(self,cb_type=None):
+        if cb_type:
+            return 'p=%s,,'%(cb_type)
+        return 'n,,'
```
Should this be `y` instead? The [RFC](https://datatracker.ietf.org/doc/html/rfc5802#section-6) says that the client must not use `n` if it supports channel binding.

I think the channel binding information should be passed to the non-Plus mechanism too, and if it's present, then this should be a `y`?

Commit:
Add SASL SCRAM-SHA-1/SHA-256 authentication with optional TLS channel binding

Implements SCRAM-SHA-1, SCRAM-SHA-1-PLUS, SCRAM-SHA-256 and SCRAM-SHA-256-PLUS
per RFC 5802 and RFC 7677. A new xmpp/scram.py module provides ScramBase with
ScramSHA1 / ScramSHA256 subclasses; SASL in auth.py delegates to them via the
SCRAM_MECHANISMS registry.

Key points:
- Nonces generated with os.urandom; DIGEST-MD5 cnonce uses secrets.token_hex
- GS2 header: 'p=<type>,,' for PLUS, 'y,,' when binding available but non-PLUS,
  'n,,' otherwise (RFC 5802 §7)
- Server signature verified on success for all SCRAM-SHA-* variants
- Automatic fallback from PLUS to non-PLUS when server rejects channel binding,
  preserving the 'y,,' GS2 flag on retry
- Nonce validated: server must extend the client nonce, not merely echo it
- bare 'except' replaced with 'except Exception' throughout
